File: Project-5/Python/Project_5.py
import numpy as np
import cv2
import matplotlib.pyplot as plt
import random
from scipy.optimize import least_squares

def SIFT(img1, img2):
    img1_gray = cv2.cvtColor(img1, cv2.COLOR_BGR2GRAY)
    img2_gray = cv2.cvtColor(img2, cv2.COLOR_BGR2GRAY)

    sift = cv2.SIFT_create()
    keypoints_1, descriptors_1 = sift.detectAndCompute(img1_gray, None)
    keypoints_2, descriptors_2 = sift.detectAndCompute(img2_gray, None)

    bf = cv2.BFMatcher(cv2.NORM_L1, crossCheck = True)
    matches = bf.match(descriptors_1, descriptors_2)
    matches = sorted(matches, key = lambda x:x.distance)

    corres = []
    for match in matches:
        p1 = keypoints_1[match.queryIdx].pt
        p2 = keypoints_2[match.trainIdx].pt
        corres.append([[*p1], [*p2]])

    # Visualize the result
    img12 = cv2.drawMatches(img1, keypoints_1, img2, keypoints_2, matches[:1000], img2, flags=2)

    return corres, img12

def Homography(corres, pairs_num):
    A = np.zeros((pairs_num*2, 8))
    b = np.zeros((pairs_num*2, 1))

    for i in range(pairs_num):
        x, y = corres[i][0]
        xp, yp = corres[i][1]

        A[i*2] = np.array([x, y, 1, 0, 0, 0, -x*xp, -y*xp])
        A[i*2+1] = np.array([0, 0, 0, x, y, 1, -x*yp, -y*yp])

        b[i*2] = xp
        b[i*2+1] = yp

    H = np.append(np.linalg.pinv(A)@b, [1]).reshape(3,3)

    return H

def Homography_RANSAC(corres, rand_num, N, M):
    domain_coord = np.array([row[0] for row in corres])
    domain_coord = np.insert(domain_coord, 2, 1, axis = 1)
    range_coord = np.array([row[1] for row in corres])
    range_coord = np.insert(range_coord, 2, 1, axis = 1)

    inliers_num = -1
    inliers_idx = []

    for i in range(N):
        random_index = random.sample(range(len(corres)), rand_num)
        
        corres_rand = np.array([corres[i] for i in random_index])
        H_initial = Homography(corres_rand, pairs_num = rand_num)

        domain_coord_proj = H_initial@domain_coord.T
        domain_coord_proj = (domain_coord_proj / domain_coord_proj[2, :]).T

        error = np.abs(domain_coord_proj - range_coord)
        d = np.sum(error**2, axis = 1)
        idx = np.where(d <= delta**2)[0]

        if len(idx) > inliers_num:
            inliers_idx = idx
            inliers_num = len(inliers_idx)
            H_ransac = H_initial

        if len(idx) > M:
            break

    return H_ransac, inliers_idx


def visualize_RANSAC(img_left, img_right, img_id, corres, inliers_idx):
    img_in = np.concatenate((img_left, img_right), axis = 1)
    img_out = np.concatenate((img_left, img_right), axis = 1)

    for i in range(len(corres)):
        left = tuple(np.round(corres[i][0]).astype(int))
        right = tuple(np.round(corres[i][1]).astype(int) + np.array([img_left.shape[1], 0]))
        if i in inliers_idx:
            cv2.circle(img_in, left, radius = 4, color = (0,255,0), thickness = -1)
            cv2.circle(img_in, right, radius = 4, color = (0,255,0), thickness = -1)
            cv2.line(img_in, left, right, (0,255,0), 1)
        else:
            cv2.circle(img_out, left, radius = 4, color = (0,0,255), thickness = -1)
            cv2.circle(img_out, right, radius = 4, color = (0,0,255), thickness = -1)
            cv2.line(img_out, left, right, (0,0,255), 1)

    # Save the output image
    newStr = "img" + str(img_id) + str(img_id+1) + '_RANSAC_inliers.jpg'
    cv2.imwrite("/Users/" + newStr, img_in)
    print(newStr, "saved.")
    newStr = "img" + str(img_id) + str(img_id+1) + '_RANSAC_outliers.jpg'
    cv2.imwrite("/Users/" + newStr, img_out)
    print(newStr, "saved.")

    return None


def Homography_LLS(corres, inliers_idx):
    corres_in = np.array([corres[i] for i in inliers_idx])
    n = len(corres_in)

    A = np.zeros((n*2, 8))
    b = np.zeros((n*2, 1))

    for i in range(n):
        x, y = corres_in[i][0]
        xp, yp = corres_in[i][1]

        A[i*2] = np.array([x, y, 1, 0, 0, 0, -x*xp, -y*xp])
        A[i*2+1] = np.array([0, 0, 0, x, y, 1, -x*yp, -y*yp])

        b[i*2] = xp
        b[i*2+1] = yp

    A_pseudo = np.linalg.pinv(A.T@A)@A.T
    h = A_pseudo@b
    H_LLS = np.append(h, [1]).reshape(3,3)

    return H_LLS


def cost_function(h, domain_coord, range_coord):
    H = h.reshape(3,3)

    domain_coord = np.insert(domain_coord, 2, 1, axis = 1)
    range_coord = np.insert(range_coord, 2, 1, axis = 1)

    domain_coord_proj = H@domain_coord.T
    domain_coord_proj = (domain_coord_proj / domain_coord_proj[2, :]).T

    error = np.abs(domain_coord_proj - range_coord)
    error_norm = (error**2)[:,:2]

    return error_norm.flatten()


def Homography_LM(correspondences, inliers_idx):
    domain_coord_in = np.array([correspondences[i][0] for i in inliers_idx])
    range_coord_in = np.array([correspondences[i][1] for i in inliers_idx])

    LM = least_squares(cost_function, H_LLS.flatten(), args=(domain_coord_in, range_coord_in), method = "lm")
    h_LM = np.array(LM.x)
    h_LM /= h_LM[-1]

    H_LM = h_LM.reshape(3,3)

    return H_LM


def Proj_to_pano(panorama_img, H, img_src):
    single_img_height, single_img_width, _ = img_src.shape
    x = np.repeat(np.arange(single_img_width), single_img_height)
    y = np.tile(np.arange(single_img_height), single_img_width)
    w = np.ones((single_img_height*single_img_width), dtype = int)
    xyw = np.column_stack((x,y,w))

    new_coord = H@xyw.T
    new_coord /= new_coord[2,:]
    panorama_coord = (new_coord[:2]).T + np.array([single_img_width*2, 0])

    # Remove the coordinates that are negative or out of frame
    valid_coord = (panorama_coord[:,1] > 0) & (panorama_coord[:,1] < single_img_height-1)
    panorama_coord_valid = np.round(panorama_coord[valid_coord]).astype(int)
    xy_valid = xyw[valid_coord][:,:2]

    # Project to the panorama image
    for pano, xy in zip(panorama_coord_valid, xy_valid):
        if np.all(panorama_img[pano[1], pano[0]] == [0,0,0]):
            if 0<xy[1]<single_img_height-1 and 0<xy[0]<single_img_width-1 and 0<pano[1]<single_img_height-1 and 0<pano[0]<single_img_width-1:
                panorama_img[pano[1]-1:pano[1]+1, pano[0]-1:pano[0]+1] = img_src[xy[1]-1:xy[1]+1, xy[0]-1:xy[0]+1]
            else:
                panorama_img[pano[1]-1:pano[1]+1, pano[0]-1:pano[0]+1] = img_src[xy[1], xy[0]]

    return None


def panorama_generate(images_num, H_refine_all):
    # Load images
    image_set = {}
    for i in range(images_num):
        image_set['img%d' %i] = cv2.imread("/Users/%d.jpg"%(i))

    # Define the panorama image
    single_img_height, single_img_width, _ = image_set['img0'].shape
    panorama_height, panorama_width = single_img_height, single_img_width*5
    panorama_img = np.zeros((panorama_height, panorama_width, 3), dtype = np.uint8)

    mid = images_num//2
    panorama_img[:, single_img_width*2:single_img_width*3] = image_set['img%d' %mid]     # Set the middle image

    # Project all images to one
    Proj_to_pano(panorama_img, H = H_refine_all[1], img_src = image_set['img1'])
    Proj_to_pano(panorama_img, H = H_refine_all[0]@H_refine_all[1], img_src = image_set['img0'])
    Proj_to_pano(panorama_img, H = np.linalg.inv(H_refine_all[2]), img_src = image_set['img3'])
    Proj_to_pano(panorama_img, H = np.linalg.inv(H_refine_all[2]@H_refine_all[3]), img_src = image_set['img4'])

    return panorama_img


if __name__ == '__main__':

    ###### Automatically estimating homographies #####
    # Outlier rejection using the RANSAC algorithm
    epsilon = 0.6
    p = 0.99
    n = 5                                                          # Number of correspondences to choose in each trial
    N = np.ceil(np.log(1-p)/np.log(1-(1-epsilon)**n)).astype(int)  # Number of trials
    delta = 5                                                      # 3*sigma (sigma is set to a small number between 0.5 and 2)

    H_LLS_all = []
    H_LM_all = []
    for i in range(4):
        img1 = cv2.imread("/Users/%d.jpg"%(i))
        img2 = cv2.imread("/Users/%d.jpg"%(i+1))

        correspondences, img_SIFT = SIFT(img1, img2)

        # Save the output image
        newStr = "img%d%d"%(i, i+1) + '_SIFT' + '.jpeg'
        cv2.imwrite("/Users/" + newStr, img_SIFT)
        print(newStr, "saved.")

        n_total = len(correspondences)                              # Total number of correspondences
        M = np.ceil((1-epsilon)*n_total).astype(int)                # Number of inliers in the data

        H_ransac, inliers_idx = Homography_RANSAC(corres = correspondences, rand_num = n, N = N, M = M)
        visualize_RANSAC(img_left = img1, img_right = img2, img_id = i, corres = correspondences, inliers_idx = inliers_idx)

        H_LLS = Homography_LLS(correspondences, inliers_idx)        # Homography estimation using a Linear Least-Squares method
        H_LM = Homography_LM(correspondences, inliers_idx)          # Homography refinement using a Nonlinear Least-Squares approach

        H_LLS_all.append(H_LLS)
        H_LM_all.append(H_LM)
    print("Task 1.4 Done")
    
    #### Project all the views on to a fixed common frame #####
    output_LLS = panorama_generate(images_num = 5, H_refine_all = H_LLS_all)
    cv2.imwrite("Panorama_image_LLS.jpg", output_LLS)
    print("Panorama_image_LLS.jpg saved.")
    output_LM = panorama_generate(images_num = 5, H_refine_all = H_LM_all)
    cv2.imwrite("Panorama_image_LM.jpg", output_LM)
    print("Panorama_image_LM.jpg saved.")

    print("Task 1.5 Done")

File: Project-5/cpp/Project_5.cpp
#include <iostream> 
#include <opencv2/opencv.hpp> 
#include <string>
#include <vector>
#include <Eigen/Dense>
#include <random>
#include <cmath>
#include <thread>

using namespace std;
using namespace cv; 
using namespace Eigen;

int num_images;
int img_width, img_height, pano_img_width;
const float epsilon = 0.6;
const float p = 0.99;
const int rand_num = 6; // Number of correspondences to choose in each trial
const int N = ceil(log(1 - p) / log(1 - pow(1 - epsilon, rand_num)));  // Number of trials
const int delta = 5; // 3*sigma (sigma is set to a small number between 0.5 and 2)
int M; // Threshold of number of inliers

cv::Mat ComputeHomography(const vector<Point2f>& points_1, const vector<Point2f>& points_2, int pairs_num){

    MatrixXd A(pairs_num * 2, 8);
    VectorXd b(pairs_num * 2);

    float x, y, xp, yp;

    for (int i = 0; i < pairs_num; i++){
        x = points_1[i].x;
        y = points_1[i].y;
        xp = points_2[i].x;
        yp = points_2[i].y;

        A.row(i*2) << x, y, 1, 0, 0, 0, -x*xp, -y*xp;
        A.row(i*2+1) << 0, 0, 0, x, y, 1, -x*yp, -y*yp;
        b(i*2, 0) = xp;
        b(i*2+1, 0) = yp;
    }

    Eigen::VectorXd h = A.jacobiSvd(Eigen::ComputeFullU | Eigen::ComputeFullV).solve(b);
    cv::Mat H = (Mat_<double>(3,3) << h(0), h(1), h(2), h(3), h(4), h(5), h(6), h(7), 1);

    return H;
}


cv::Mat ComputeHomography_LLS(const vector<Point2f>& points_1, const vector<Point2f>& points_2){

    int pairs_num = points_1.size();
    MatrixXd A(pairs_num * 2, 8);
    VectorXd b(pairs_num * 2);

    float x, y, xp, yp;
    RowVectorXd newRow(8);

    for (int i = 0; i < pairs_num; i++){
        x = points_1[i].x;
        y = points_1[i].y;
        xp = points_2[i].x;
        yp = points_2[i].y;

        A.row(i*2) << x, y, 1, 0, 0, 0, -x*xp, -y*xp;
        A.row(i*2+1) << 0, 0, 0, x, y, 1, -x*yp, -y*yp;
        b(i*2, 0) = xp;
        b(i*2+1, 0) = yp;
    }

    VectorXd h = (A.transpose() * A).ldlt().solve(A.transpose() * b);
    cv::Mat H_LLS = (Mat_<double>(3,3) << h(0), h(1), h(2), h(3), h(4), h(5), h(6), h(7), 1);

    return H_LLS;
}


cv::Mat getRefinedHomographyFromImagePair(const Mat img1, const Mat img2){

    // Convert images to grayscale
    Mat img1_gray, img2_gray;
    cvtColor(img1, img1_gray, COLOR_BGR2GRAY);
    cvtColor(img2, img2_gray, COLOR_BGR2GRAY);

    // SIFT feature detector and descriptor
    Ptr<SIFT> detector = SIFT::create();
    vector<KeyPoint> keypoints_1, keypoints_2;
    Mat descriptors_1, descriptors_2;

    detector -> detectAndCompute(img1_gray, noArray(), keypoints_1, descriptors_1);
    detector -> detectAndCompute(img2_gray, noArray(), keypoints_2, descriptors_2);

    // BFMatcher with default params
    BFMatcher matcher(NORM_L1, true);
    vector<DMatch> matches;
    matcher.match(descriptors_1, descriptors_2, matches);
    int n_matches = matches.size();

    // Sort matches by distance
    sort(matches.begin(), matches.end(), [](const DMatch& a, const DMatch& b) {
        return a.distance < b.distance;
    });

    // Store coordinates
    vector<Point2f> matchpoints_1, matchpoints_2;
    for (const auto& match : matches) {
        matchpoints_1.push_back(keypoints_1[match.queryIdx].pt);
        matchpoints_2.push_back(keypoints_2[match.trainIdx].pt);
    }

    // Reject outliers with RANSAC
    M = static_cast<int>(ceil((1 - epsilon) * n_matches)); // Threshold of number of inliers

    int inliers_num = 0;
    vector<Point2f> matchpoints_1_RANSAC, matchpoints_2_RANSAC;
    random_device rd;
    mt19937 g(rd());
    for (int i = 0; i < N; ++i) { 
        vector<Point2f> matchpoints_1_random, matchpoints_2_random;

        // Random sampling
        vector<int> indices(n_matches);
        iota(indices.begin(), indices.end(), 0);
        shuffle(indices.begin(), indices.end(), g);
        for (int j = 0; j < rand_num; ++j) {
            matchpoints_1_random.push_back({matchpoints_1[indices[j]]});
            matchpoints_2_random.push_back({matchpoints_2[indices[j]]});
        }

        // Compute initial homography
        cv::Mat H_initial = ComputeHomography(matchpoints_1_random, matchpoints_2_random, rand_num);

        // Find inliers
        vector<int> current_inliers;
        std::vector<Point2f> projected_points(n_matches);
        cv::perspectiveTransform(matchpoints_1, projected_points, H_initial);
        
        vector<Point2f> matchpoints_1_inliers, matchpoints_2_inliers;
        for (int i = 0; i < n_matches; i++){
            double error = norm(matchpoints_2[i] - projected_points[i]);
            if (error <= pow(delta, 2)) {
                matchpoints_1_inliers.push_back(matchpoints_1[i]);
                matchpoints_2_inliers.push_back(matchpoints_2[i]);
            }
        }

        if (matchpoints_1_inliers.size() > inliers_num) {
            inliers_num = matchpoints_1_inliers.size();
            matchpoints_1_RANSAC = matchpoints_1_inliers;
            matchpoints_2_RANSAC = matchpoints_2_inliers;
        }

        if (current_inliers.size() > M) {
            break;
        }
    }

    Mat H_LLS = ComputeHomography_LLS(matchpoints_1_RANSAC, matchpoints_2_RANSAC);

    return H_LLS;
}


void projectOneImage(cv::Mat pano_img, const cv::Mat src_img, const cv::Mat H, const bool debug){

    // Project img to the panorama image
    std::vector<Point2f> obj_pixel(1);
    std::vector<Point2f> scene_pixel(1);
    Point2f offset = {static_cast<float>(img_width*((num_images-1)/2)), 0};

    for (int x = 0; x < img_width; x++){ 
        for (int y = 0; y < img_height; y++){ 
            
            obj_pixel[0] = cv::Point(x, y);
            
            perspectiveTransform(obj_pixel, scene_pixel, H);
            scene_pixel[0] = scene_pixel[0] + offset;
            int scene_pixel_x = round(scene_pixel[0].x);
            int scene_pixel_y = round(scene_pixel[0].y);

            if ((scene_pixel_x < pano_img_width) && (scene_pixel_y < img_height) && (pano_img.at<cv::Vec3b>(scene_pixel_y, scene_pixel_x) == cv::Vec3b(0, 0, 0))){
                if ((0 <= scene_pixel_x) && (scene_pixel_x < pano_img_width) && (0 <= scene_pixel_y) && (scene_pixel_y < img_height)) {
                    if (debug) cout << "obj_pixel = " << obj_pixel << ", scene_pixel = [" << scene_pixel_x << ", " << scene_pixel_y << "]\n";
                    pano_img.at<cv::Vec3b>(scene_pixel_y, scene_pixel_x) = src_img.at<cv::Vec3b>(y, x);
                }
                // fill the holes
                if ((0 < scene_pixel_x) && (scene_pixel_x < pano_img_width-1) && (0 < scene_pixel_y) && (scene_pixel_y < img_height-1)){
                    pano_img.at<cv::Vec3b>(scene_pixel_y+1, scene_pixel_x) = src_img.at<cv::Vec3b>(y, x);
                    pano_img.at<cv::Vec3b>(scene_pixel_y-1, scene_pixel_x) = src_img.at<cv::Vec3b>(y, x);
                    pano_img.at<cv::Vec3b>(scene_pixel_y, scene_pixel_x+1) = src_img.at<cv::Vec3b>(y, x);
                    pano_img.at<cv::Vec3b>(scene_pixel_y, scene_pixel_x-1) = src_img.at<cv::Vec3b>(y, x);
                }
            }
        }
    }
    cout << "\nProject image done.\n";
}


int main(){

    cout << "\nStarting Project 5 ... \n";
    num_images = 5;

    // Read in images
    Mat img0 = imread("0.jpg");
    Mat img1 = imread("1.jpg");
    Mat img2 = imread("2.jpg");
    Mat img3 = imread("3.jpg");
    Mat img4 = imread("4.jpg");

    if (img1.empty()) {
        std::cerr << "Image 1 not found" << std::endl;
        return -1;
    }

    img_width = img1.size().width;   // 747
    img_height = img1.size().height; // 1328
    pano_img_width = img_width * num_images; // 2241

    // Declare homography matrices
    cv::Mat H_01, H_12, H_23, H_34;

    // Launch threads for each homography computation
    std::thread t1([&]{ H_01 = getRefinedHomographyFromImagePair(img0, img1); });
    std::thread t2([&]{ H_12 = getRefinedHomographyFromImagePair(img1, img2); });
    std::thread t3([&]{ H_23 = getRefinedHomographyFromImagePair(img2, img3); });
    std::thread t4([&]{ H_34 = getRefinedHomographyFromImagePair(img3, img4); });

    t1.join();
    t2.join();
    t3.join();
    t4.join();

    // Create a new image with the total width of the two images
    Mat pano_img(img_height, pano_img_width, CV_8UC3, cv::Scalar(0, 0, 0));

    // Place middle image
    img2.copyTo(pano_img(Rect(img_width*int((num_images)/2), 0, img_width, img_height)));
    cout << "\nMiddle img copy to middle\n";

    // Project images
    projectOneImage(pano_img, img1, H_12, false);
    projectOneImage(pano_img, img0, H_01 * H_12, false);
    projectOneImage(pano_img, img3, H_23.inv(), false);
    projectOneImage(pano_img, img4, (H_23 * H_34).inv(), false);

    cv::imwrite("Pano_image.png", pano_img);
    
    return 0;
}
